R/options.R

```diff
@@ -1,270 +1,100 @@
 #' @examples
-#' pl$set_polars_options(strictly_immutable = FALSE)
-#' pl$get_polars_options()
+#' pl$set_options(maintain_order = TRUE, strictly_immutable = FALSE)
+#' pl$options
 #'
-#'
-#' # setting strictly_immutable = 42 will be rejected as
+#' # these options only accept booleans (TRUE or FALSE)
 #' tryCatch(
-#'   pl$set_polars_options(strictly_immutable = 42),
+#'   pl$set_options(strictly_immutable = 42),
 #'   error = function(e) print(e)
 #' )
 #'
-pl$set_polars_options = function(
-    ...,
-    return_replaced_options = TRUE) {
-  # check opts
-  opts = list2(...)
-  if (is.null(names(opts)) || !all(nzchar(names(opts)))) stopf("all options passed must be named")
-  unknown_opts = setdiff(names(opts), names(polars_optenv))
-  if (length(unknown_opts)) {
-    stopf(paste("unknown option(s) was passed:", paste(unknown_opts, collapse = ", ")))
-  }
-
-  # update options
-  replaced_opts_list = list()
-  for (i in names(opts)) {
-    opt_requirements = polars_optreq[[i]]
-    stopifnot(
-      !is.null(opt_requirements),
-      is.list(opt_requirements),
-      all(sapply(opt_requirements, is.function)),
-      all(nzchar(names(opt_requirements)))
-    )
-
-    for (j in names(opt_requirements)) {
-      opt_check = opt_requirements[[j]]
-      opt_value = opts[[i]]
-      opt_result = opt_check(opt_value)
-      if (!isTRUE(opt_result)) {
-        stopf(paste(
-          "option:", i, " must satisfy requirement named", j,
-          "with function\n", paste(capture.output(print(opt_check)), collapse = "\n")
-        ))
-      }
+#' # reset options to their default value
+#' pl$reset_options()
+
+pl$set_options = function(
+    strictly_immutable = TRUE,
+    maintain_order = FALSE,
+    do_not_repeat_call = FALSE,
+    debug_polars = FALSE,
+    no_messages = FALSE
+  ) {
+
+  # only modify arguments that were explicitly written in the function call
+  # (otherwise calling set_options() twice in a row would reset the args
+  # modified in the first call)
+  args_modified = names(as.list(sys.call()[-1]))
+  for (i in seq_along(args_modified)) {
+    value = get(args_modified[i])
+    if (length(value) > 1) {
+      stop(paste0("`", args_modified[i], "` must be of length 1."))
     }
-
-    replaced_opts_list[[i]] = polars_optenv[[i]]
-    polars_optenv[[i]] = opts[[i]]
-  }
-
-  if (return_replaced_options) {
-    return(invisible(replaced_opts_list))
-  }
-
-  # return current option set invisible
-  invisible(pl$get_polars_options())
-}
-
-
-
-polars_opts_defaults = as.list(polars_optenv)
-#' @rdname polars_options
-#' @name reset_polars_options
-#' @examples
-#' # reset options like this
-#' pl$reset_polars_options()
-pl$reset_polars_options = function() {
-  rm(list = ls(envir = polars_optenv), envir = polars_optenv)
-  for (i in names(polars_opts_defaults)) {
-    polars_optenv[[i]] = polars_opts_defaults[[i]]
+    if (!is.logical(value)) {
+      stop(paste0("`", args_modified[i], "` only accepts `TRUE` or `FALSE`."))
```

Comment:
On a sidenote:

[I tried an experiment with direct active binding for rust rpool get/set ](https://github.com/pola-rs/r-polars/commit/73d129e4ecbcad42138529c0ff7525618c803dae)

Works in principle, except I think there is a bug in R when a set-binding function is run from a locked environment. "R gets confused" and think it is an assignment and raises an the error below, which for assigning to a locked environment, but it is not an assigment

anyways this should not be a problem for `pl$set_options`

```R
> pl$binding_opts$rpool_cap
$available
[1] 0

$capacity
[1] 5

> pl$binding_opts$rpool_cap <- 25
Error in pl$binding_opts$rpool_cap <- 25 : 
  cannot change value of locked binding for 'binding_opts'
> pl$binding_opts$rpool_cap
$available
[1] 0

$capacity
[1] 25

> 
```

Reply:
suggestion of how to add rust side options https://github.com/pola-rs/r-polars/commit/252fee9869e90e57f30ba1a272f8ddbe91a1efd9

i did not get it quote right but maybe for inspiration, otherwise I will try again later

but in generel im positive of this PR as it is, and nice to haves could be added later

Reply:
now it works, small bug , so this could be how to add rust side options 76dd5884ee20c272661ba5cf4786847720133ddd

Reply:
Cool, but I'd rather address that in another PR later (building on your branch) so that this one doesn't wait for too long